src/modes/ModeEstimation.jl

```diff
@@ -440,3 +442,190 @@ function _optimize(
+# Generic optimisation objective initialisation #
+#################################################
+
+function (f::Turing.OptimLogDensity)(G, z)
+  spl = DynamicPPL.SampleFromPrior()
+  
+  # Calculate log joint and the gradient
+  l, g = Turing.gradient_logp(
+      z, 
+      DynamicPPL.VarInfo(f.vi, spl, z), 
+      f.model, 
+      spl,
+      f.context
+  )
+
+  # Use the negative gradient because we are minimizing.
+  G[:] = -g
+
+  return nothing
+end
+
+
+
+function transform!(f::OptimLogDensity)
+  spl = DynamicPPL.SampleFromPrior()
+
+  ## Check link status of vi in OptimLogDensity
+  linked = DynamicPPL.islinked(f.vi, spl) 
+
+  ## transform into constrained or unconstrained space depending on current state of vi
+  if !linked
+    f.vi[spl] = f.vi[DynamicPPL.SampleFromPrior()]
+    DynamicPPL.link!(f.vi, spl)
+  else
+    DynamicPPL.invlink!(f.vi, spl)
+  end
+
+  return nothing
+end
+
+
+function transform2constrained(par::AbstractArray, vi::DynamicPPL.VarInfo)
+  spl = DynamicPPL.SampleFromPrior()
+
+  linked = DynamicPPL.islinked(vi, spl)
+  
+  !linked && DynamicPPL.link!(vi, spl)
+  vi[spl] = par
+  DynamicPPL.invlink!(vi,spl)
+  tpar = vi[spl]
+  
+  linked && DynamicPPL.link!(vi,spl)
+
+  return tpar
+end
+
+function transform2constrained!(par::AbstractArray, vi::DynamicPPL.VarInfo)
+  par .= transform2constrained(par, vi)
+  return nothing
+end
+
+function transform2unconstrained(par::AbstractArray, vi::DynamicPPL.VarInfo)
+  spl = DynamicPPL.SampleFromPrior()
+
+  linked = DynamicPPL.islinked(vi, spl)
+  linked && DynamicPPL.invlink!(vi, spl)
+  vi[spl] = par
+  DynamicPPL.link!(vi, spl)
+  tpar = vi[spl]
+  !linked && DynamicPPL.invlink!(vi, spl)
+
+  return tpar
+end
+
+function transform2unconstrained!(par::AbstractArray, vi::DynamicPPL.VarInfo)
+  par .= transform2unconstrained(par, vi)
+  return nothing
+end
+
+
+function _orderpar(par::NamedTuple, vi::DynamicPPL.VarInfo)
+  tmp_idx = indexin(collect(keys(vi.metadata)), collect(keys(par)))
+  tmp = collect(par)[tmp_idx]
+  return tmp, tmp_idx
+end
+
+
+function transform2unconstrained(par::NamedTuple, vi::DynamicPPL.VarInfo; order::Bool=true, array::Bool=true)
+  par_sor, par_idx = _orderpar(par, vi)
+
+  transform2unconstrained!(par_sor, vi)
+
+  if order
+    if array
+      return par_sor
+    else
+      return (; zip(keys(vi.metadata), par_sor)...)
+    end
+  else
+    if array
+      return par_sor[tmp_idx]
+    else
+      return (; zip(keys(par),par_sor[par_idx])...)
+    end
+  end
+end
+
+function transform2constrained(par::NamedTuple, vi::DynamicPPL.VarInfo; order::Bool=true, array::Bool=true)
+  par_sor, par_idx = _orderpar(par, vi)
+
+  transform2constrained!(par_sor, vi)
+
+  if order
+    if array
+      return par_sor
+    else
+      return (; zip(keys(vi.metadata), par_sor)...)
+    end
+  else
+    if array
+      return par_sor[tmp_idx]
+    else
+      return (; zip(keys(par),par_sor[par_idx])...)
+    end
+  end
+end
+
```

Comment:
IMO we should not implement this here (at least not for now). It feels like we should add better support for using NamedTuples in DynamicPPL instead (also for specifying initial values, outputs, etc.).
```suggestion
```

Reply:
Makes sense. I will remove this